src/user/pipes/username-validation.pipe.ts

```diff
@@ -0,0 +1,24 @@
+import { Injectable, PipeTransform } from '@nestjs/common'
+
+import { ConflictError } from '../../common/errors/errors'
+import { UserDatabaseService } from '../user.database.service'
+import type { CreateUserDto } from '../dto/create-user.dto'
+
+@Injectable()
+export class UsernameValidationPipe implements PipeTransform {
```

Comment:
Interesting use of pipes 🧡
However, I suggest moving this logic to the UserService: pipes are good for data validation/transformation; but handling conflicts it's responsibility of business-logic layer.

Reply:
Also, you're jumping over the layer, from Controller layer directly to the DB layer, which just highlights this architectural flaw.

Reply:
Yeah, seems like I misunderstood a bit the core concept of pipes 😵
Fixed, thanks!